### PR TITLE
bpo-47172: Make virtual opcodes negative. Make is_jump detect only actual jumps, use is_block_push for the exception block setup opcodes.

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue47172

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -157,6 +166,7 @@ write_instr(_Py_CODEUNIT *codestr, struct instr *instruction, int ilen)
             *codestr++ = _Py_MAKECODEUNIT(EXTENDED_ARG, (oparg >> 8) & 0xFF);
             /* fall through */
         case 1:
+            assert(!IS_VIRTUAL_OPCODE(opcode));
```
Comment (brandtbucher) [inline Python/compile.c]: I'd prefer to move this just after we grab `opcode` on line 155. A virtual opcode would likely cause strange problems before we get to this point (for example, when looking up the cache count).

File: Python/compile.c [hunk 2]
```diff
@@ -7593,6 +7596,7 @@ assemble_jump_offsets(struct assembler *a, struct compiler *c)
             bsize = b->b_offset;
             for (i = 0; i < b->b_iused; i++) {
                 struct instr *instr = &b->b_instr[i];
+                assert(!IS_VIRTUAL_OPCODE(instr->i_opcode));
```
Comment (brandtbucher) [inline Python/compile.c]: I think, conceptually, this sort of belongs in `instr_size`. Not a big deal, though.

File: Python/compile.c [hunk 3]
```diff
@@ -9183,13 +9189,6 @@ duplicate_exits_without_lineno(struct compiler *c)
      */
     for (basicblock *b = c->u->u_blocks; b != NULL; b = b->b_list) {
         if (b->b_iused > 0 && is_jump(&b->b_instr[b->b_iused-1])) {
-            switch (b->b_instr[b->b_iused-1].i_opcode) {
-                /* Note: Only actual jumps, not exception handlers */
-                case SETUP_WITH:
-                case SETUP_FINALLY:
-                case SETUP_CLEANUP:
-                    continue;
-            }
```
Comment (brandtbucher) [inline Python/compile.c:9214]: Just curious, why is this removed?

File: Python/compile.c [hunk 4]
```diff
@@ -8615,7 +8620,7 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
         struct instr *inst = &bb->b_instr[i];
         int oparg = inst->i_oparg;
         int nextop = i+1 < bb->b_iused ? bb->b_instr[i+1].i_opcode : 0;
-        if (is_jump(inst)) {
+        if (is_jump(inst) || is_block_push(inst)) {
```
Comment (brandtbucher) [inline Python/compile.c:8650]: Hm, I didn't realize virtual opcodes made it this far.

Is there any value to including them in the jump threading prep here, though? The peephole optimizer doesn't know how to handle them anyways...

**Review by brandtbucher — APPROVED** [review]
Nice. I have a few questions/suggestions:

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c:9214]: ``is_jump(&b->b_instr[b->b_iused-1])``   before this PR returns True for the SETUP_X opcodes, but now it doesn't so we don't need to exclude them anymore.

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: instr_size is called from other places, where this might not be valid?

File: Python/compile.c (hunk 4 above)
Comment (iritkatriel) [reply Python/compile.c:8650]: Yes, this is very early after codegen, and before the exception table is constructed.

This loop jumps over empty blocks at the target of a jump, I think it (theoretically) could happen with a setup-X.

File: Python/compile.c (hunk 2 above)
Comment (brandtbucher) [reply Python/compile.c]: Pretty sure it's fine? I don't think it's ever possible to actually compute the size of virtual instructions; even computing the cache requirements is illegal!

File: Python/compile.c (hunk 4 above)
Comment (brandtbucher) [reply Python/compile.c:8650]: Ah, disregard my comment. I forgot that we were actually updating `inst->i_target` here!

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: I think you're right.

**Review by brandtbucher — APPROVED** [review]

File: Python/compile.c [hunk 5]
```diff
@@ -124,16 +126,24 @@ is_relative_jump(struct instr *i)
...
+{
+    int opcode = instr->i_opcode;
+    return opcode == SETUP_FINALLY || opcode == SETUP_WITH || opcode == SETUP_CLEANUP;
+}
+
 static inline int
 is_jump(struct instr *i)
 {
-    return i->i_opcode >= SETUP_WITH || is_bit_set_in_table(_PyOpcode_Jump, i->i_opcode);
+    return is_bit_set_in_table(_PyOpcode_Jump, i->i_opcode);
```
Comment (brandtbucher) [inline Python/compile.c]: Ah wait, I think calling `is_bit_set_in_table` on virtual opcodes is illegal. Probably still need a guard here?

```suggestion
    return !IS_VIRTUAL_OPCODE(i->i_opcode) && is_bit_set_in_table(_PyOpcode_Jump, i->i_opcode);
```

**Review by raghavthind2005 — APPROVED** [review]

File: Python/compile.c [hunk 6]
```diff
@@ -72,18 +72,20 @@
 
 /* Pseudo-instructions used in the compiler,
  * but turned into NOPs by the assembler. */
-#define SETUP_FINALLY 255
-#define SETUP_CLEANUP 254
-#define SETUP_WITH 253
-#define POP_BLOCK 252
-#define JUMP 251
+#define SETUP_FINALLY 256
```
Comment (markshannon) [inline Python/compile.c]: Just an idea, but would it make sense for the virtual opcodes to be negative?
`>0`: Real
`==0`: Cache
`<0`: Virtual 
It seems like a simpler scheme.

File: Python/compile.c [hunk 7]
```diff
@@ -72,18 +72,20 @@
...
-#define SETUP_CLEANUP 254
-#define SETUP_WITH 253
-#define POP_BLOCK 252
-#define JUMP 251
+#define SETUP_FINALLY -1
+#define SETUP_CLEANUP -2
+#define SETUP_WITH -3
+#define POP_BLOCK -4
+#define JUMP -5
+
```
Comment (markshannon) [inline Python/compile.c:80]: Could you add a `MINIMUM_VIRTUAL_OPCODE` or similar and add an assert when creating an instruction that 
`MINIMUM_VIRTUAL_OPCODE <= opcode && opcode <= MAXIMUM_ALLOWED_OPCODE`

`MAXIMUM_ALLOWED_OPCODE` is currently 254, as we use 255 for tracing.

File: Python/compile.c [hunk 8]
```diff
@@ -115,8 +117,13 @@ is_bit_set_in_table(const uint32_t *table, int bitindex) {
      * Word is indexed by (bitindex>>ln(size of int in bits)).
      * Bit within word is the low bits of bitindex.
      */
-    uint32_t word = table[bitindex >> LOG_BITS_PER_INT];
-    return (word >> (bitindex & MASK_LOW_LOG_BITS)) & 1;
+    if (bitindex >= 0) {
```
Comment (markshannon) [inline Python/compile.c]: Might as well be extra safe and add `bitindex < 256`.

**Review by markshannon — COMMENTED** [review]
Looks good. A couple of small suggestions

**Comment by markshannon** [comment]
Thanks
